dynamite_nsm/services/kibana/package/manager.py

```diff
@@ -0,0 +1,784 @@
+import tarfile
+import json
+import re
+import os
+import requests
+from getpass import getpass
+from collections import defaultdict
+from tabulate import tabulate
+from uuid import uuid4
+from unidecode import unidecode
+from io import BytesIO, IOBase
+from typing import AnyStr, Optional, Tuple, IO, List
+from dynamite_nsm.logger import get_logger
+from dynamite_nsm.utilities import get_primary_ip_address
+from dynamite_nsm.utilities import PrintDecorations as PD
+from dynamite_nsm.services.kibana.package.mappings import PACKAGES_INDEX_MAPPING, PACKAGES_INDEX_NAME
+from dynamite_nsm.services.kibana.package.schemas import ORPHAN_OBJECT_PACKAGE_MANIFEST_DATA as OrphanPackageData, \
+                                                         InstalledObjectSchema, \
+                                                         PackageManifestSchema, \
+                                                         SchemaToObject
+
+class InstalledObject(SchemaToObject):
+    def __init__(self, json_data):
+        super().__init__(json_data, InstalledObjectSchema())
+
+    @staticmethod
+    def from_kwargs(**kwargs):
+        data = {}
+        data['title'] = kwargs.get('title')
+        data['object_type'] = kwargs.get('object_type', None)
+        data['object_id'] = kwargs.get('object_id', None)
+        data['space_id'] = kwargs.get('space_id', None)
+        obj = InstalledObject(data)
+        return obj
+
+    @staticmethod
+    def from_installation_result(responsedata, **kwargs):
+        data = {}
+        data['title'] = responsedata.get('meta', {'title': 'untitled object'}).get('title')
+        data['object_type'] = responsedata.get('type', None)
+        # this is the id of the saved object, not the document for tracking
+        data['object_id'] = responsedata.get('id', None)
+        data.update(kwargs)
+        obj = InstalledObject.from_kwargs(**data)
+        return obj
+
+    def __repr__(self) -> str:
+        return f"<InstalledObect id: {self.object_id}, title: {self.title} >"
+
+class PackageManifest(SchemaToObject):
+    def __init__(self, json_data):
+        """
+        :param json_data: JSON Body of the package manifest, conforms to PackageManifestSchema
+        """
+        super().__init__(json_data, PackageManifestSchema())
+        self.data['slug'] = self.create_slug()
+
+    def create_slug(self):
+        name = self.name or OrphanPackageData.get('name')
+        # using unidecode to support unicode/ascii in the package manifest data.
+        slug = unidecode(name).lower()
+        return re.sub(r'[\W_]+', '-', slug)
+    
+    def json(self) -> str:
+        if not self.data.get('slug', None):
+            self.data['slug'] = self.slug
+        return json.dumps(self.data)
+
+    def __repr__(self) -> str:
+        return f"<PackageManifest(name={self.name}, author={self.author})>"
+
+class Package(object):
+
+    package_index_name = PACKAGES_INDEX_NAME
+    es_url = f'https://{get_primary_ip_address()}:9200'
```

Comment:
We can't always assume this. Totally possible that ES doesn't live local to the device the package manager lives on. We also cannot assume the port number. This is also configurable.

Considering kibana package manager can be used as a standalone application we cannot even always rely on retreiving the right target from the configuration files.


Here's the logic I'd apply to the situation.

Check if Kibana is installed locally. (kibana.profile.ProfileManager().is_installed()).
- if it is installed, grab the location of the host/port via the`kibana.install.ConfigManager().elasticsearch_targets` (Use index 0).
- If it is not installed, force the user to provide the target-string as an argument.

Reply:
Update made to use kibana API to proxy request to ES

Reply:
Further changes made per discussion during scrum. Now provides a `target` option for the user to supply the url to kibana (validated with urlparser), as well as preferring `host:port` from kibana config data pulled from dynamite environment file, and finally falling back to `primary_ip:5601` if all else fails.